### PR TITLE
Changes for reading LEF files

### PR DESCRIPTION
**Pull request by jtsylve**
This rather large PR adds improved support for handling the reading of LEF (L01) files

### CODE REVIEW

**Comment by jtsylve** [comment]
We're looking into the test_write_functions CI failure

**Comment by joachimmetz** [comment]
@jtsylve can you be more specific what theses changes fix? Also what is the relationship between this PR and the 2 PRs closes without a comment?

**Comment by jtsylve** [comment]
If you need more specifics, I'll have to ask the person who worked on this to comment, but basically this PR enhances libewf's support for LEF files in a number of ways.

1) It exposes new functionality to the API for querying a number of new metadata properties of stored logical files

2) It fixes issues with reading certain standards compliant LEF files that read fine in Encase and FTK, but seem to break tools that rely on libewf.

The two PRs that I closed were part of my failed attempt at breaking this larger PR into several smaller PRs, that can be ignored.

**Comment by kastonework** [comment]
Detail on the changes: The majority of the changes are additions to libewf to support reading of additional single file entry data from LEFs (guid, fileID, deletedDate, record type, sources tree, subjects tree, permissions tree, and extended attributes tree). 

The other change was a fix for a bug identified in libewf:
* Fix issue where streamed data was truncated if it was not a multiple of the ev file sector size (could be reproduced via ewfacquirestream).

There had been another bug fix in there, but that was causing the CI test failure so it was reverted. That issue (missing number_of_chunks and number_of_sectors in data section of ev files with >64 segments) will likely get reported as a bug with more detail.

**Comment by joachimmetz** [comment]
@kastonework @jtsylve seeing this PR changes a lot of code and differs from the code style it will take significant time to get the changes merged. And my time is very limited at the moment.

> It fixes issues with reading certain standards compliant LEF files that read fine in Encase and FTK, but seem to break tools that rely on libewf.

Can you provide more detail or test files. It is important that differences in the data format are also documented. Since the "standard" is proprietary.

**Comment by kastonework** [comment]
> ... and differs from the code style it will take significant time to get the changes merged. 

The intention was to match the existing code style; if you can point out an example of differences then I will update the new code to match.

**Comment by jtsylve** [comment]
@joachimmetz Is there a clang-format (or similar tool) style that can be applied to automatically format the code for consistency?

**Comment by joachimmetz** [comment]
> Is there a clang-format (or similar tool) style that can be applied to automatically format the code for consistency?

unfortunately not, but not worries, I mainly mentioned this to indicate that merging will take some time.

What I do would like to know is:

> It fixes issues with reading certain standards compliant LEF files that read fine in Encase and FTK, > but seem to break tools that rely on libewf.

Can you provide more detail or test files. It is important that differences in the data format are also documented. Since the "standard" is proprietary.

**Comment by joachimmetz** [comment]
Leaving comments FYI and notes to self while preparing the code for merge.

File: common/config_winapi.h [hunk 1]
```diff
@@ -50,11 +50,11 @@
 
 /* Enable verbose output
 #define HAVE_VERBOSE_OUTPUT		1
- */
```
Comment (joachimmetz) [inline common/config_winapi.h]: Changes not needed

File: include/libewf.h.in [hunk 2]
```diff
@@ -1499,6 +1499,673 @@ int libewf_handle_get_file_entry_by_utf16_path(
      libewf_file_entry_t **file_entry,
      libewf_error_t **error );
 
+/* Retrieves the (single) file source for the specific id
```
Comment (joachimmetz) [inline include/libewf.h.in]: use identifier instead of id

File: include/libewf.h.in [hunk 3]
```diff
@@ -1499,6 +1499,673 @@ int libewf_handle_get_file_entry_by_utf16_path(
...
+
+/* Frees a file source
+ * Returns 1 if successful or -1 on error
+ */
+LIBEWF_EXTERN \
+int libewf_file_source_free(
+	libewf_file_source_t **file_source,
+	libewf_error_t **error);
+ 
+LIBEWF_EXTERN \
```
Comment (joachimmetz) [inline include/libewf.h.in]: missing docstings

File: include/libewf.h.in [hunk 4]
```diff
@@ -1499,6 +1499,673 @@ int libewf_handle_get_file_entry_by_utf16_path(
...
+	 size_t *utf16_mac_address_size,
+	 libewf_error_t **error);
+
+LIBEWF_EXTERN \
+int libewf_file_source_get_utf16_mac_address(
+	 libewf_file_source_t *file_source,
+	 uint16_t *utf16_mac_address,
+	 size_t utf16_mac_address_size,
+	 libewf_error_t **error);
+
```
Comment (joachimmetz) [inline include/libewf.h.in]: additional empty line not necessary, also trailing white space was introduced

File: include/libewf.h.in [hunk 5]
```diff
@@ -1499,6 +1499,673 @@ int libewf_handle_get_file_entry_by_utf16_path(
...
+ * Subject functions
+ * ------------------------------------------------------------------------- */
+
+/* Frees a file subject
+ * Returns 1 if successful or -1 on error
+ */
+LIBEWF_EXTERN \
+int libewf_file_subject_free(
+	libewf_file_subject_t **file_subject,
+	libewf_error_t **error);
```
Comment (joachimmetz) [inline include/libewf.h.in]: space missing between `error` and `);`

File: include/libewf.h.in [hunk 6]
```diff
@@ -1499,6 +1499,673 @@ int libewf_handle_get_file_entry_by_utf16_path(
...
+/* -------------------------------------------------------------------------
+ * Subject functions
+ * ------------------------------------------------------------------------- */
+
+/* Frees a file subject
+ * Returns 1 if successful or -1 on error
+ */
+LIBEWF_EXTERN \
+int libewf_file_subject_free(
+	libewf_file_subject_t **file_subject,
```
Comment (joachimmetz) [inline include/libewf.h.in]: tab was used for indentation instead of corresponding number of spaces

File: libewf/libewf_handle.c [hunk 7]
```diff
@@ -6463,7 +6498,7 @@ ssize_t libewf_internal_handle_write_finalize_file_io_pool(
 		 */
 		internal_handle->media_values->number_of_chunks  = internal_handle->write_io_handle->number_of_chunks_written;
 		internal_handle->media_values->number_of_sectors = (uint64_t) ( internal_handle->write_io_handle->input_write_count / internal_handle->media_values->bytes_per_sector );
-		internal_handle->media_values->media_size        = (size64_t) internal_handle->write_io_handle->input_write_count;
+		internal_handle->media_values->media_size        = (size64_t)internal_handle->write_io_handle->input_write_count - (size64_t)internal_handle->write_io_handle->input_write_sector_padding;
```
Comment (joachimmetz) [inline libewf/libewf_handle.c]: PR contains multiple type of changes, detangle these

File: libewf/libewf_single_file_extended_attribute.h [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	 size_t *utf16_string_size,
+	 libcerror_error_t **error);
+
+int libewf_single_file_extended_attribute_get_utf16_value(
+	 libewf_single_file_extended_attribute_t *single_file_extended_attribute,
+	 uint16_t *utf16_string,
+	 size_t utf16_string_size,
+	 libcerror_error_t **error);
+
+#pragma pack(push, 1)
```
Comment (joachimmetz) [inline libewf/libewf_single_file_extended_attribute.h]: why is this defined half way through the source file, move to separate source file.

File: libewf/libewf_single_file_extended_attribute.h [hunk 9]
```diff
@@ -0,0 +1,153 @@
+/*
+ * Single file source functions
+ *
```
Comment (joachimmetz) [inline libewf/libewf_single_file_extended_attribute.h]: missing license notice, see: https://www.gnu.org/licenses/gpl-howto.html

File: libewf/libewf_single_file_extended_attribute.h [hunk 10]
```diff
@@ -0,0 +1,153 @@
+/*
+ * Single file source functions
```
Comment (joachimmetz) [inline libewf/libewf_single_file_extended_attribute.h]: source => extended attributes

File: libewf/libewf_single_file_extended_attribute.h [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+	uint32_t level;
+	uint8_t folder;
+	uint32_t name_size;
+	uint32_t value_size;
+};
+#pragma pack(pop)	
+
+int libewf_single_file_decode_extended_attribute_parse_item(
+	libewf_single_file_extended_attribute_t *ea_item,
+	uint8_t **bufCur,
```
Comment (joachimmetz) [inline libewf/libewf_single_file_extended_attribute.h]: deviation of style

File: libewf/libewf_single_file_extended_attribute.h [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+{
+	uint32_t level;
+	uint8_t folder;
+	uint32_t name_size;
+	uint32_t value_size;
+};
+#pragma pack(pop)	
+
+int libewf_single_file_decode_extended_attribute_parse_item(
+	libewf_single_file_extended_attribute_t *ea_item,
```
Comment (joachimmetz) [inline libewf/libewf_single_file_extended_attribute.h]: use of abbreviations

File: msvscpp/bzip2/bzip2.vcproj [hunk 13]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="Windows-1252"?>
 <VisualStudioProject
 	ProjectType="Visual C++"
-	Version="9,00"
+	Version="9.00"
```
Comment (joachimmetz) [inline msvscpp/bzip2/bzip2.vcproj]: these files are auto generated.

File: tests/Makefile.am [hunk 14]
```diff
@@ -532,6 +533,18 @@ ewf_test_single_file_entry_LDADD = \
 	../libewf/libewf.la \
 	@LIBCERROR_LIBADD@
 
+ewf_test_single_file_source_SOURCES = \
```
Comment (joachimmetz) [inline tests/Makefile.am]: no tests for extended attribute?

File: libewf/libewf_single_file_extended_attribute_tree.c [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+#include "libewf_libcdata.h"
+#include "libewf_libcerror.h"
+#include "libewf_libuna.h"
+#include "libewf_single_file_extended_attribute.h"
+#include "libewf_single_file_extended_attribute_tree.h"
+
+/* Retrieves the single file extended_attribute sub node for the specific extended_attribute id
+ * Returns 1 if successful, 0 if no such sub single file extended_attribute or -1 on error
+ */
+int libewf_single_file_extended_attribute_tree_get_sub_node_by_id(
```
Comment (joachimmetz) [inline libewf/libewf_single_file_extended_attribute_tree.c]: is this function used?

**Comment by joachimmetz** [comment]
@jtsylve to better understand this large PR and to ensure there is test coverage and not to merge unused code I'll make similar changes based on PR but not necessarily the same.

File: libewf/libewf_single_files.c [hunk 16]
```diff
@@ -1499,53 +4439,47 @@ int libewf_single_files_parse_file_entry(
...
+					&value_64bit,
+					64,
+					LIBFVALUE_INTEGER_FORMAT_TYPE_DECIMAL_UNSIGNED,
+					error) != 1)
+				{
+					libcerror_error_set(
+						error,
+						LIBCERROR_ERROR_DOMAIN_MEMORY,
+						LIBCERROR_MEMORY_ERROR_SET_FAILED,
+						"%s: unable to set size.",
```
Comment (joachimmetz) [inline libewf/libewf_single_files.c]: "unable to set size" => "unable to set file identifier"

File: libewf/libewf_single_files.c [hunk 17]
```diff
@@ -1604,6 +4538,26 @@ int libewf_single_files_parse_file_entry(
 			else if( ( type_string[ 0 ] == (uint8_t) 'p' )
 			      && ( type_string[ 1 ] == (uint8_t) 'm' ) )
 			{
+				if (value_string_size > 0 && *value_string == '-')
```
Comment (joachimmetz) [inline libewf/libewf_single_files.c]: To my knowledge only `-1` is a known exception value. @jtsylve do you have insight into more exception values?

File: libewf/libewf_single_files.c [hunk 18]
```diff
@@ -988,50 +1533,2314 @@ int libewf_single_files_parse_record_values(
...
+			/* Ignore empty values
+			 */
+		}
+		else if (type_string_size == 4)
+		{
+			if ((type_string[0] == (uint8_t) 'p')
+			 && (type_string[1] == (uint8_t) 'g')
+			 && (type_string[2] == (uint8_t) 'u'))
+			{
+				single_file_source->primary_guid = (uint8_t *)memory_allocate(
```
Comment (joachimmetz) [inline libewf/libewf_single_files.c]: @jtsylve are these GUIDs little or big endian? do you also have a sample file / sample section that contains a source.

File: libewf/libewf_single_files.c [hunk 19]
```diff
@@ -988,50 +1533,2314 @@ int libewf_single_files_parse_record_values(
...
+						"%s: unable to set name.",
+						function);
+
+					goto on_error;
+				}
+				single_file_source->name[value_string_size - 1] = 0;
+
+				single_file_source->name_size = value_string_size;
+			}
+			else if (type_string[0] == (uint8_t) 'p')
```
Comment (joachimmetz) [inline libewf/libewf_single_files.c]: @jtsylve this does not make sense in the context of source, was this just copied from file entry?

File: libewf/libewf_single_file_permission.h [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+	 */
+	uint8_t *name;
+
+	/* The name size
+	 */
+	size_t name_size;
+
+	/* The uid
+	 */
+	uint8_t *uid;
```
Comment (joachimmetz) [inline libewf/libewf_single_file_permission.h]: @jtsylve could you confirm that this value is not only used for a user identifier (uid) but also for a group identifier (gid) if the (permission) type is not owner?

File: libewf/libewf_single_file_permission.h [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+	 */
+	size_t uid_size;
+	
+	/* The props
+	 */
+	uint32_t props;
+
+	/* The access mask
+	 */
+	uint32_t access_mask;
```
Comment (joachimmetz) [inline libewf/libewf_single_file_permission.h]: @jtsylve aren't these the [Windows NT Access control entry (ACE) flags](https://github.com/libyal/libfwnt/blob/master/documentation/Security%20Descriptor.asciidoc#access-control-entry-ace-flags)?

Comment (joachimmetz) [reply libewf/libewf_single_file_permission.h]: Do you have sample data you can share?

File: libewf/libewf_single_files.h [hunk 22]
```diff
@@ -59,6 +59,18 @@ struct libewf_single_files
 	/* The single file entry tree
 	 */
 	libcdata_tree_node_t *root_file_entry_node;
+
+	/* The single file source tree
+	 */
+	libcdata_tree_node_t *root_file_source_node;
```
Comment (joachimmetz) [inline libewf/libewf_single_files.h]: these might be stored as "trees", they appear to be more like an arrays

File: libewf/libewf_single_files.c (hunk 19 above)
Comment (joachimmetz) [reply libewf/libewf_single_files.c]: I disagree with how `'p'` is interpreted here, I've updated the documentation.

File: libewf/libewf_single_file_permission.h (hunk 20 above)
Comment (joachimmetz) [reply libewf/libewf_single_file_permission.h]: I was able to confirm this myself, uid is is not the most fitting name. I'll go for (ACE) identifier.

File: libewf/libewf_single_file_permission.h (hunk 21 above)
Comment (joachimmetz) [reply libewf/libewf_single_file_permission.h]: I was able to confirm this myself, updated the documentation.

**Comment by joachimmetz** [comment]
@jtsylve @kastonework @mikebbt if one of you has sample files with actual LEF sources and subjects (not the defaults ones) that would be helpful. I do not necessarily need the whole file, just the ltree (single files) serialized object data (string).

File: libewf/libewf_single_file_source.c [hunk 23]
```diff
@@ -0,0 +1,4116 @@
...
+		 function );
+
+		return( -1 );
+	}
+	if( utf8_string_size < single_file_source->sha1_hash_size )
+	{
+		libcerror_error_set(
+		 error,
+		 LIBCERROR_ERROR_DOMAIN_ARGUMENTS,
+		 LIBC
```
Comment (joachimmetz) [inline libewf/libewf_single_file_source.c]: This is a hexstring where "000000000000" represents not set.

File: libewf/libewf_single_files.c [hunk 24]
```diff
@@ -988,50 +1533,2314 @@ int libewf_single_files_parse_record_values(
...
+					libcerror_error_set(
+						error,
+						LIBCERROR_ERROR_DOMAIN_MEMORY,
+						LIBCERROR_MEMORY_ERROR_INSUFFICIENT,
+						"%s: unable to create mac_address.",
+						function);
+
+					goto on_error;
+				}
+				if (narrow_string_copy(
```
Comment (joachimmetz) [inline libewf/libewf_single_files.c]: This is a hexstring where "000000000000" represents not set.

File: libewf/libewf_single_files.c [hunk 25]
```diff
@@ -988,50 +1533,2314 @@ int libewf_single_files_parse_record_values(
...
+						"%s: unable to create GUID.",
+						function);
+
+					goto on_error;
+				}
+
+				if (libewf_single_files_read_hex_string(
+					value_string,
+					value_string_size,
+					single_file_source->guid,
```
Comment (joachimmetz) [inline libewf/libewf_single_files.c]: encase refers to this a device guid

File: libewf/libewf_single_files.c (hunk 18 above)
Comment (joachimmetz) [reply libewf/libewf_single_files.c]: I found a sample file, but if you have closure on the byte order let me know.

**Comment by jtsylve** [comment]
I didn't actually do the work in these patches.  I'll ask @kastonework to respond.

**Comment by joachimmetz** [comment]
Ack, then I should credit her instead.

**Comment by kastonework** [comment]
@joachimmetz @jtsylve I’ll review and respond to all of the questions in the next day or so.

**Comment by joachimmetz** [comment]
@kastonework no hurries, I'll keep working on adding equivalent functionality. I'll ask you to double check if everything is there when I'm finished. I'm deviating from your original approach in some places. 

What could be useful in the intermediate time is having specific test data:

* single files with multiple sources
* single files one or more subjects

**Comment by jtsylve** [comment]
> Ack, then I should credit her instead.

Yes, Please credit Kim and BlackBag for the work.  I just put in the PR.

File: libewf/libewf_single_files.c (hunk 17 above)
Comment (kastonework) [reply libewf/libewf_single_files.c]: Yes - as far as I'm aware, -1 is the only value that we'd expect for an item that doesn't have permissions.

File: libewf/libewf_single_files.h (hunk 22 above)
Comment (kastonework) [reply libewf/libewf_single_files.h]: True - there will only be one level to the tree so it's essentially an array.

**Comment by kastonework** [comment]
> @jtsylve @kastonework @mikebbt if one of you has sample files with multiple LEF sources and/or actual subjects (not the defaults ones) that would be helpful. I do not necessarily need the whole file, just the ltree (single files) serialized object data (string).

I can provide this; should I email it to you?

File: libewf/libewf_single_files.c (hunk 25 above)
Comment (kastonework) [reply libewf/libewf_single_files.c]: Yeah - it would be clearer to refer to it as such rather than as a more generic "guid"

**Comment by joachimmetz** [comment]
> I can provide this; should I email it to you?

@kastonework yes please

File: libewf/libewf_single_files.c (hunk 24 above)
Comment (kastonework) [reply libewf/libewf_single_files.c]: Yes, I see that in sample logical ev files as well.

**Comment by joachimmetz** [comment]
was auto closed due to migration to main branch, will look into merging parts of the PR that will be beneficial

**Pull request closed without merging**